src/calendar.cpp

```diff
@@ -556,11 +556,24 @@ time_duration rng( time_duration lo, time_duration hi )
 
 season_type season_of_year( const time_point &p )
 {
-    if( calendar::eternal_season() ) {
-        // If we use calendar::start to determine the initial season, and the user shortens the season length
-        // mid-game, the result could be the wrong season!
-        return calendar::initial_season;
+    static time_point prev_turn = calendar::before_time_starts;
+    static season_type prev_season = SPRING;
+    static bool is_eternal = false;
+    
+    if( p != prev_turn ) {
+        prev_turn = p;
+        if( ( is_eternal = calendar::eternal_season() ) ) {
+            // If we use calendar::start to determine the initial season, and the user shortens the season length
+            // mid-game, the result could be the wrong season!
+            return calendar::initial_season;
+        }
+        const int season = to_turn<int>( p ) / to_turns<int>( calendar::season_length() );
+        return prev_season = static_cast<season_type>( season % 4 );
+    } else {
+        if( is_eternal ) {
```

Comment:
`is_eternal` here will never be true, since it is set in the `if`

Reply:
if ( p != prev_turn ) else if ( is_eternal )

Not sure what you're getting at.

Reply:
What happens when `p == prev_turn`, where does `is_eternal` get assigned?

Reply:
It'll have been set when p != prev_turn.

Reply:
I guess you're also caching the call to `calendar::eternal_season()`, so it makes sense in that regard.

It still seems weird that there's such a huge performance hit from checking world options.  Maybe the caching belongs at the options level, and then it's easy to bust the cache when the options change.

Reply:
I initialized is_eternal to calendar::eternal_season(), as well. Seems the best default value.

The performance hit is just because of the sheer number of times it's called. Like, on Tiles, fully lit and zoomed out with pixel minimap on, this function gets called roughly 30,000 times per turn, leading to reading the options twice, adding up to roughly 60,000 calls to get_option(). And every one of those 59,999 calls returns the same exact thing as the first call did.

Reply:
And, I agree that a means of handling cases like this should happen at the options level. I just don't have any bright ideas on how to go about it.

Reply:
OK, how about this.  If `eternal_seasion()` cache the `calendar::initial_season` value in `prev_season`.  That eliminates the need for the `is_eternal` variable and the entire `else` block

Reply:
This look acceptable?